modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt

```diff
@@ -25,39 +20,20 @@ import kotlinx.coroutines.launch
 class UpsellViewModel @Inject constructor(
     private val analyticsTracker: AnalyticsTrackerWrapper,
     private val subscriptionManager: SubscriptionManager,
-    feature: FeatureWrapper,
-    private val releaseVersion: ReleaseVersionWrapper,
 ) : ViewModel() {
 
     private val _state: MutableStateFlow<UiState> = MutableStateFlow(UiState.Loading)
     val state: StateFlow<UiState> = _state
 
     init {
-        val bookmarksFeature = feature.bookmarksFeature
-        val patronExclusiveAccessRelease = (bookmarksFeature.tier as? FeatureTier.Plus)?.patronExclusiveAccessRelease
-
-        val isReleaseCandidate = releaseVersion.currentReleaseVersion.releaseCandidate != null
-        val relativeToEarlyPatronAccess = patronExclusiveAccessRelease?.let {
-            releaseVersion.currentReleaseVersion.comparedToEarlyPatronAccess(it)
-        }
-        val availableForFeatureTier = when (relativeToEarlyPatronAccess) {
-            null -> bookmarksFeature.tier
-            EarlyAccessState.Before,
-            EarlyAccessState.During,
-            -> if (isReleaseCandidate) bookmarksFeature.tier else FeatureTier.Patron
-
-            EarlyAccessState.After -> bookmarksFeature.tier
-        }
-        val subscriptionTier = availableForFeatureTier.toSubscriptionTier()
-
         viewModelScope.launch {
-            subscriptionManager.freeTrialForSubscriptionTierFlow(subscriptionTier)
```

Comment:
I had the same feeling about removing this. My initial thought was that since we use GitHub, we could simply copy and paste this logic from there as needed. However, I agree that this isn't the most convenient approach.

What just occurred to me is that we could extract this logic into a separate module. This way, we can easily remove or add it as needed. But I should create an issue for this since the current PR is already quite large. What do you think?

I can also keep this logic here to ensure we don't lose it

Reply:
Ok, so I've created the `modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/earlyaccess/EarlyAccessHelper.kt` to keep this logic and unit tests for it. Let me know what do you think